cookiecutter/vcs.py

```diff
@@ -115,6 +187,10 @@ def clone(repo_url, checkout=None, clone_to_dir='.', no_input=False):
                     'The {} branch of repository {} could not found, '
                     'have you made a typo?'.format(checkout, repo_url)
                 )
-            raise
+            raise RepositoryCloneFailed(
```

Comment:
Hey @Theta-Dev, I've just found out that can be a breaking change because no one traps the bare `subprocess.CalledProcessError`. This is useful to get the full traceback on an unhandled exception. What do you think about it?

Reply:
What do you mean with "no one traps the bare CalledProcessError"?

If a known error is encountered, it outputs the respective error message. Otherwise it outputs a basic error message together with the raw error output of the VCS command.

```python
try:
    vcs.clone(repo_url, checkout, clone_to_dir, repo_dir)
except subprocess.CalledProcessError as clone_error:
    output = clone_error.output.decode('utf-8')
    if 'not found' in output.lower():
        raise RepositoryNotFound(
            'The repository {} could not be found, '
            'have you made a typo?'.format(repo_url)
        )
    if any(error in output for error in BRANCH_ERRORS):
        raise RepositoryCloneFailed(
            'The {} branch of repository {} could not found, '
            'have you made a typo?'.format(checkout, repo_url)
        )
    raise RepositoryCloneFailed(
        'Cloning of Repository {} returned an error:\n{}'.format(
            repo_url, output
        )
    )
```

Just like this:

```
(venv) thetadev@ThetaBook:~/Documents/Python/cookiecutter$ cookiecutter svn+https://xsvn.code.sf.net/p/cookiecutter-pypackage/code/trunk/cookiecutter-pypackages
Cloning of Repository https://xsvn.code.sf.net/p/cookiecutter-pypackage/code/trunk/cookiecutter-pypackages returned an error:
svn: E170013: Unable to connect to a repository at URL 'https://xsvn.code.sf.net/p/cookiecutter-pypackage/code/trunk/cookiecutter-pypackages'
svn: E670005: No address associated with hostname
```
Yes, the original version did output the full stack trace in this case. But is that really more debug and user friendly?

```
thetadev@ThetaBook:~$ cookiecutter kit@github.com:audreyfeldroy/cookiecutter-pypackage
Traceback (most recent call last):
  File "/usr/local/bin/cookiecutter", line 8, in <module>
    sys.exit(main())
  File "/usr/lib/python3/dist-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/cookiecutter/cli.py", line 140, in main
    cookiecutter(
  File "/usr/local/lib/python3.8/dist-packages/cookiecutter/main.py", line 67, in cookiecutter
    repo_dir, cleanup = determine_repo_dir(
  File "/usr/local/lib/python3.8/dist-packages/cookiecutter/repository.py", line 110, in determine_repo_dir
    cloned_repo = clone(
  File "/usr/local/lib/python3.8/dist-packages/cookiecutter/vcs.py", line 101, in clone
    subprocess.check_output(
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['git', 'clone', 'kit@github.com:audreyfeldroy/cookiecutter-pypackage']' returned non-zero exit status 128.
```
By raising the bare exception you wont see the error message from the VCS and have a hard time debugging it (if it's not obviouse, like this typo).
Plus it makes it seem like there is a problem with Cookiecutter itself, which is not the case.

Reply:
@Theta-Dev, thanks for your answer.

> Yes, the original version did output the full stack trace in this case. But is that really more debug and user friendly?

Of course not, but you should consider that a cookiecutter is a high used package; this means that someone can already expect the `subprocess.CalledProcessError` exception and do something when it happens. If we change this behaviour, we can break things.

What do you think about it?